### ISSUE TITLE
Create Products/Offfers API - INTERNAL_SERVER_ERROR

### ISSUE THREAD

**Issue by aswiniimpro** [allegro/allegro-api#4685]
/sale/product-offers gets "Internal server error" message.

{ "product": { "name": "xxx", "category":{ "id":"89060" }, "parameters":[ { "name": "EAN", "values": [ "xxx" ] }, { "id": "209878", "values": [ "xxx" ] } ], "images":[ "xxx" ] }, "sellingMode": { "price": { "amount": "xxx", "currency": "PLN" } }, "stock": { "available": xxx } }

{
    "errors": [
        {
            "code": "INTERNAL_SERVER_ERROR",
            "message": "Internal server error.",
            "details": null,
            "path": null,
            "userMessage": "Internal server error. Please try again in a few minutes."
        }
    ]
}

**Comment by MaciejFrackowiak** [comment]
Please, give us a trace-id of your request - we will be able to check it in our logs.

**Comment by aswiniimpro** [comment]
Sorry, I didn't noted it's trace-id.

**Comment by MaciejFrackowiak** [comment]
If you do not have a trace-id, you can always send us your request and response via [the contact form](https://allegro.pl/help/contact?category=b469d1bc-4353-4e65-a65b-6cac33aa595b&subjectId=517305f2-5ea1-4b3b-9080-d4cc960ab6a1) - we will check them. In your message, please add GitHub thread number - 4685.

**Comment by aswiniimpro** [comment]
Ok.

When I try to create product in live /sale/product-offers, I got CATEGORY_MISMATCH error.

**Request** : 
{
  "product": {
    "name":"XXX",
    "category":{
        "id":"11889"
     },
     "images": [
"XXX"
],
          "parameters":[{
                "name": "EAN",
                "values": [
                    "0064547232478"
                ]
            }
        ]
},
"name":"XXX",
  "sellingMode": {
    "price": {
      "amount": "XXX",
      "currency": "PLN"
    }
  },
  "stock": {
    "available": 1
  },
  "external": {
	"id": "XXX"
  },
"publication": {
    "status": "INACTIVE",
    "startingAt": null,
    "startedAt": null,
    "endingAt": null,
    "endedAt": null
}
}

**Response**:

{
    "errors": [
        {
            "code": "CATEGORY_MISMATCH",
            "message": "Existing Product related to submitted data was found. The provided category `Części i akcesoria` (4588) does not match the existing product category `Rolki, prowadnice, zębatki` (11900)",
            "details": null,
            "path": "product.category.id",
            "userMessage": "Existing Product related to submitted data was found. The provided category `Części i akcesoria` (4588) does not match the existing product category `Rolki, prowadnice, zębatki` (11900)."
        }
    ]
}

**Comment by MaciejFrackowiak** [comment]
That error message means that we found a product in our database which has similar parameters - and its category is different from what you provided. That product id=ca9564da-9857-4f9c-ab2d-367a901b5742 and it has assigned category=11900.

**Comment by aswiniimpro** [comment]
Thanks. 
But, when I search product for this ean https://api.allegro.pl/sale/products?ean=0064547232478, it shows empty  "products": []  - https://prnt.sc/19pk8zj

In product id=ca9564da-9857-4f9c-ab2d-367a901b5742 - EAN is 064547232478. The ean we have is 0064547232478 (having 0 prefixed to the existed EAN)

When create products /sale/product-offers using ean 0064547232478 , It shows error 'category_mismatch' . Then, how can we confirm if it is already existed.

**Comment by MaciejFrackowiak** [comment]
Our system recognizes EAN=064547232478 as the same as EAN=0064547232478. So it is able to browse the database and find products as in the example with product.id=ca9564da-9857-4f9c-ab2d-367a901b5742.
You as a seller can report any lack of EAN code in a product by https://allegro.pl/offer/product/report/{productId}.

**Comment by aswiniimpro** [comment]
Thanks .

I've created a product using https://api.allegro.pl/sale/product-offers. I got response with product id : null. And, the product is created in ACTIVE status in allegro dashboard.
This product does not existed in Allegro when check EAN 0740013099058.

{
    "id": "10943517763",
    "name": "XXXX",
    "product": {
        "id": null,
        "publication": {
            "status": "NOT_LISTED"
        }
    },
    "afterSalesServices": {
        "impliedWarranty": {
            "id": "XXX"
        },
        "returnPolicy": {
            "id": "XXXXf"
        },
        "warranty": null
    },
    "payments": {
        "invoice": "VAT"
    },
    "sellingMode": {
        "format": "BUY_NOW",
        "price": {
            "amount": XXX,
            "currency": "PLN"
        },
        "startingPrice": null,
        "minimalPrice": null
    },
    "stock": {
        "available": 1,
        "unit": "UNIT"
    },
    "location": {
        "countryCode": "XXX",
        "province": XXX,
        "city": "XXX",
        "postCode": "XXX"
    },
    "delivery": {
        "shippingRates": {
            "id": "2e61029a-0c34-454b-b1b0-98969b234df6"
        },
        "handlingTime": "PT24H",
        "additionalInfo": null
    },
    "publication": {
        "duration": null,
        "status": "INACTIVE",
        "endedBy": null,
        "endingAt": null,
        "startingAt": null,
        "republish": false
    },
    "description": {
        "sections": [
            {
                "items": [
                    {
                        "type": "TEXT",
                        "content": "<p>XXX</p>"
                    }
                ]
            }
        ]
    },
    "validation": {
        "errors": [],
        "warnings": [],
        "validatedAt": "2021-07-22T15:09:33.349Z"
    },
    "createdAt": "2021-07-22T15:09:33Z",
    "updatedAt": "2021-07-22T15:09:33.349Z",
    "images": [
        "https://a.allegroimg.com/original/..../...."
    ],
    "external": {
        "id": "XXX"
    },
    "category": {
        "id": "260341"
    }
}

Please help us to solve this issue.

**Comment by MaciejFrackowiak** [comment]
We checked your request - as a parameter 17589 you passed an ambiguous value - this parameter identifies a product, so its value could not be ambiguous.

**Comment by aswiniimpro** [comment]
We have sent parameters as following,
"parameters": [
            {
                "name": "EAN",
                "values": [
                    "0740013099058"
                ]
            },
            {
                "id": 17589,
                "values": [],
                "valuesIds": [
                    "17589_64"
                ]
            }
]

Our item's color does not matches with the option given in "17589" parameter. So, we sent with this "17589_64 - Other" ( { "id": "17589_64", "value": "inny", "dependsOnValueIds": [] } )

Offer created, but product id is null. 

How could we send the product having 'color' that does not matched with given values in allegro ?

**Comment by PrzemyslawLukanowski** [comment]
In your case you should choose value "17589_1" (black) - even in the offer title and description you have this information - "H11-6321-220 Canon z tonerem Black" and "CANON FX-2 TONER BLACK". In this category, it is not possible to create products with an ambiguous value in this parameter.

**Comment by aswiniimpro** [comment]
How could we send the product having 'color' that does not matched with given option values in allegro parameters "17589" ?

**Comment by PrzemyslawLukanowski** [comment]
We support creating products only with the listed values in this parameter. In the case above you should choose "17589_1" (black) value, because this is the suitable value for the given product.

**Comment by aswiniimpro** [comment]
Thanks.

But my question is if some product have tricolor or 'color' that does not matched with given option values in allegro parameters "17589". What value can I choose?

**Example** : My product color is "Brown" and product-name is "S-2745 Riso FR-2950 Ink Brown"
In this, what value will be suitable to choose?

Please help us to resolve this issue.

**Comment by PrzemyslawLukanowski** [comment]
We have added "17589_935979" (brown) value in this parameter, so now you can create this product. When it comes to tricolor, you can use "17589_16" (set).

**Comment by aswiniimpro** [comment]
Ok Thank you.

what is the use of "Other - 17589_64" which is a "ambiguousValueId": "17589_64" in color parameter in category 260341? Why can't I select Other and create product?

Because, We have few more different colors like Green, Violet, Orange, Purple which is not listed in the color option values.

**Comment by PrzemyslawLukanowski** [comment]
You can still use "Other - 17589_64" when you create an offer that is not connected to a product.

I sent your suggestions to our team, I will let you know when we add those values.

**Comment by aswiniimpro** [comment]
Ok Thanks.

**Comment by PrzemyslawLukanowski** [comment]
We've added the mentioned values to the "Color" parameter.

**Comment by aswiniimpro** [comment]
Thanks for adding.

We have few more different combination of colors. What we can do or choose for the following types of color or what value will be suitable/taken for creating the product 

Sepia, Black / Color, CMY, Tri-Color, Silver, CMYK, Brown, Gold, Chromatic Red, Light Cyan, Metallic Silver, Orange, Gloss, Medium Blue, Light Magenta, Marine Red

**Comment by PrzemyslawLukanowski** [comment]
For some of the listed colors you can use the values that already exist:
```
orange 17589_936244
black 17589_1
brown 17589_935979
Tri-Color, CMY, CMYK 17589_16 (set)
```
I am not sure "Gloss" is a color of a printer toner? Do you have a sample of this product?

For the rest of the values, I sent suggestions to our team. I will come back with an answer.

**Comment by aswiniimpro** [comment]
Thanks.

Example product for Gloss - https://www.swingdesign.com/products/roland-versauv-ink-500cc-gloss-euv4-5gl 

What about the rest of the color values?

**Comment by PrzemyslawLukanowski** [comment]
Thank you for the example. I'm still waiting for the decision of our team.

**Comment by aswiniimpro** [comment]
Any updates for the rest of the color values ?
Sepia, Black / Color,Silver, Gold, Chromatic Red, Light Cyan, Metallic Silver, Gloss, Medium Blue, Light Magenta, Marine Red

**Comment by Lukasz-Zurek** [comment]
@aswiniimpro Yes. We have just added new color values:

sepia: 17589_939774
gold: 17589_939775
silver: 17589_939776
gloss: 17589_939777
metallic silver: 17589_939778
marine red: 17589_939779
chromatic red: 17589_939780
light magenta: 17589_939781
light cyan: 17589_939782
medium blue: 17589_939783

**Comment by aswiniimpro** [comment]
Thanks for adding.

But these new colors are not displaying in sandbox https://api.allegro.pl.allegrosandbox.pl/sale/categories/260341/parameters

**Comment by Lukasz-Zurek** [comment]
@aswiniimpro Yes, the category and parameter tree on Sandbox differs from the production tree. We are making updates every now and then, trying to synchronize this data. Our last update has been [on July 8th, 2021](https://developer.allegro.pl/en/news/#sandbox-july-synchro). Please be patient. Obviously we will make these changes soon.

**Comment by aswiniimpro** [comment]
Dear Team is there any update for us?

**Comment by Lukasz-Zurek** [comment]
@aswiniimpro We do not know the exact date of the next update but as soon as we know more - we will inform about it in [our news](https://developer.allegro.pl/en/news/). Please be patient.

**Comment by aswiniimpro** [comment]
Ok.

We have tried to create an offer for EAN 4960999869353 under 260345 category. But, it shows error "CATEGORY_MISMATCH" as it was assigned in Cartridges and Inks category id (101347)  
101347 = Business and Services - Office and advertising - Stationery - Accessories.

But, we need to send this offer in Inks - Genuine category id (260345) 
260345 =  Electronics - Computers - Printers and Scanners.

Is there a way to solve this issue, we need our items under 260345

**Comment by Lukasz-Zurek** [comment]
Please note the fact that the product which you received using GET /sale/products?ean=4960999869353 - has got some other **category.similar.id**, so it looks like this:
```

...
            "category": {
                "id": "101347",
                "similar": [
                    {
                        "id": "260837"
                    },
                    {
                        "id": "64567"
                    },
                    {
                        "id": "64505"
                    },
                    {
                        "id": "101348"
                    }
                ]
            },
...
```

 It means that you will not be able to create an offer for EAN 4960999869353 under 260345 category.

But I am already consulting this case. Please be patient. I will come back to you with some more information.

**Comment by Lukasz-Zurek** [comment]
@aswiniimpro Unfortunately you will not be able to create an offer for EAN 4960999869353 under 260345 category, because it is a differrent category and we cannot add it to similar categories in 101347. Technically we cannot make any change in this situation.

**Comment by aswiniimpro** [comment]
Thank you.

Dear Team is there any update for us for color?

**Comment by MaciejFrackowiak** [comment]
@aswiniimpro Not yet - I am still waiting for a decision and reply.

**Comment by MaciejFrackowiak** [comment]
@aswiniimpro We switched the category id to 260341 for a product "2de653c1-5282-496b-8fc1-05db47ec8eb7".

**Comment by aswiniimpro** [comment]
Hi,

Invoice is generating for B2C Order and getting invoice_id in response in https://allegro.pl.allegrosandbox.pl [allegro sandbox]

Sample trace id - 43ab18366268f9f8 , 61dfa6e2e67baff4

invoice will be accept to B2B order only? I didn't get error when create invoice to B2C order in sandbox.

**Comment by MaciejFrackowiak** [comment]
@aswiniimpro Yes, you are able to add an invoice to any order, but we will not display it to the buyer, who did not want an invoice.

**Comment by aswiniimpro** [comment]
> @aswiniimpro Not yet - I am still waiting for a decision and reply.

Hi,
Any updates to us for color?

**Comment by MaciejFrackowiak** [comment]
@aswiniimpro Yes, I see that decision was taken. For example CMY, CMYK, Varnish are not the colours, so they decided not to add a new one to the list of values.
